src/cholesky/lkjcholesky.jl

```diff
@@ -0,0 +1,183 @@
+        η > 0 || throw(ArgumentError("Shape parameter must be positive."))
+    end
+    logc0 = lkj_logc0(d, η)
+    uplo = _as_char(_uplo)
+    uplo ∈ ('U', 'L') || throw(ArgumentError("uplo must be 'U' or 'L'."))
+    T = Base.promote_eltype(η, logc0)
+    LKJCholesky{T, typeof(d)}(d, T(η), uplo, T(logc0))
+end
+
+_as_char(c::Char) = c
+_as_char(x) = only(string(x))
+
+#  -----------------------------------------------------------------------------
+#  REPL display
+#  -----------------------------------------------------------------------------
+
+show(io::IO, d::LKJCholesky) = show_multline(io, d, [(:d, d.d), (:η, d.η), (:uplo, d.uplo)])
+
+#  -----------------------------------------------------------------------------
+#  Conversion
+#  -----------------------------------------------------------------------------
+
+function convert(::Type{LKJCholesky{T}}, d::LKJCholesky) where T <: Real
+    return LKJCholesky{T, typeof(d.d)}(d.d, T(d.η), d.uplo, T(d.logc0))
+end
+function convert(::Type{LKJCholesky{T}}, d::Integer, η, logc0) where T <: Real
+    return LKJCholesky{T, typeof(d)}(d, T(η), d.uplo, T(logc0))
+end
+
+#  -----------------------------------------------------------------------------
+#  Properties
+#  -----------------------------------------------------------------------------
+
+dim(d::LKJCholesky) = d.d
+
+function size(d::LKJCholesky)
+    p = dim(d)
+    return (p, p)
+end
+
+function insupport(d::LKJCholesky, R::Cholesky)
+    p = dim(d)
+    factors = R.factors
+    (isreal(factors) && size(factors, 1) == p) || return false
+    iinds, jinds = axes(factors)
+    T = typeof(one(eltype(R))^2)
+    # check that the diagonal of U'*U or L*L' is all ones
+    @inbounds if R.uplo === 'U'
+        for j in 1:p
+            s = zero(T)
+            for i in 1:j
+                s += factors[iinds[i], jinds[j]]^2
+            end
+            isapprox(s, 1) || return false
+        end
+    else  # R.uplo === 'L'
+        for i in 1:p
+            s = zero(T)
+            for j in 1:i
+                s += factors[iinds[i], jinds[j]]^2
+            end
+            isapprox(s, 1) || return false
+        end
+    end
+    return true
+end
+
+function mode(d::LKJCholesky)
+    p = dim(d)
+    factors = Matrix{partype(d)}(I, p, p)
+    return Cholesky(factors, d.uplo, 0)
+end
+
+params(d::LKJCholesky) = (d.d, d.η, d.uplo)
+
+@inline partype(::LKJCholesky{T}) where {T <: Real} = T
+
+#  -----------------------------------------------------------------------------
+#  Evaluation
+#  -----------------------------------------------------------------------------
+
+function logkernel(d::LKJCholesky, R::Cholesky)
+    factors = R.factors
+    p, η = params(d)
+    c = p + 2(η - 1)
+    T = typeof(one(c) * log(one(eltype(factors))))
+    logp = zero(T)
+    di = diagind(factors)
+    for i in 2:p
+        logp += (c - i) * log(factors[di[i]])
+    end
+    return logp
+end
+
+logpdf(d::LKJCholesky, R::Cholesky) = logkernel(d, R) + d.logc0
```

Comment:
Probably one should check `insupport(R)` at some point?

Reply:
Point against that are 1. none of the matrix distributions explicitly check `insupport` during the kernel. 2. for `d=4`, `insupport` has 30% the runtime of `logpdf`. But as `d` increases, `insupport` becomes more expensive (`insupport` is more expensive than `logpdf` around `d=20`).

Reply:
The other matrix distributions at least check the dimensions (using `logpdf` with checks and `_logpdf` without checks) and I'd argue that they should also check the support and return `logpdf(d, x) = -Inf` it `x` is not in the support, as done for uni- and matrixvariate distributions.

Reply:
> none of the matrix distributions explicitly check `insupport` during the kernel.

That's more a historical artifact than a design choice. I agree that `insupport` ought to be checked, but also that you should somehow be able to switch that off for performance purposes.

> I'd argue that they should also check the support and return `logpdf(d, x) = -Inf` it `x` is not in the support, as done for uni- and matrixvariate distributions.

Should it be `-Inf` or an error? Like, how should "isn't posdef" be handled versus "isn't the right dims"?

Reply:
> Should it be -Inf or an error? Like, how should "isn't posdef" be handled versus "isn't the right dims"?

This is a design choice but for multivariate dimensions a `DimensionMismatch` is thrown if the dimension is incorrect (e.g., https://github.com/JuliaStats/Distributions.jl/blob/d285de5e0df4290905c0eab5ed0c94ecd1037567/src/multivariates.jl#L199-L200) and `-Inf` is returned if `x` is of the correct dimension but not in the support (e.g., in the case of `Dirichlet`: https://github.com/JuliaStats/Distributions.jl/blob/d285de5e0df4290905c0eab5ed0c94ecd1037567/src/multivariate/dirichlet.jl#L142-L144). This is slightly inconsistent though since `insupport(d::Dirichlet, x)` also checks the dimension of `x`.

Reply:
Right, so is that how you want the matrix distributions to behave as well?

Reply:
I've added an `insupport` check to `logpdf` that causes an error to be raised. A user can call `_logpdf` to sidestep this.